### DISCUSSION TITLE
Interface equivalence

### DISCUSSION THREAD

**Discussion by paulhasell** [dotnet/csharplang#5540] [Language Ideas]
I'm using shared projects in visual studio to author classes that I need in multiple applications/libraries, typically this is shared data model classes shared between a Dataverse plugin assembly and one (or more) Azure hosted services. This works fine as the classes are constrained to the app in which they are compiled and I'm not attempting to cast from one to the other. However, I have fallen over a use case where that would be useful...

I have written an app that dynamically loads a Dataverse assembly and extracts information from the plugin classes to register them in Dataverse. I'm testing for the implementation of an Interface to determine which exposed classes require registration, however that Interface is defined in a shared project, so the versions in the loaded assembly and the hosting app are NOT the same and you can't cast them.

Where this any other scenario I would extract the Interface into a separate dll and reference that in both places ensuring they are actually the same thing. However, Dataverse assemblies do not support the inclusion of external dll's, thus the use of a shared source project.

So, what would be useful is support for Interface equivalence to allow me to cast based on the shared project.

I would expect equivalence to operate somewhat the same as the Tuple equivalence (but more difficult to implement I expect) such that if the interface 'signature' is identical then the cast is allowed. It could also allow for the fact that the fully qualified type name for the interface is the same in both instances.

An alternative would be to provide a specific cast variation that coerces where the author 'knows' the interfaces are identical.

This would be similar to use cases available in Typescript etc where equivalence of this sort is available.

**Comment by HaloFour** [comment]
Can you use some other method to detect the plugins by reflection, like the name of the interface or an attribute?  Interface equivalence is not something that C# could support without explicit runtime support, and this scenario sounds pretty niche to be a motivating use case to consider it.

**Comment by iam3yal** [comment]
You can do what @HaloFour said or you can make sure the class you need to instantiate simply implements a certain method or multiple ones at runtime and throw if it doesn't this way you don't need to cast to anything you simply use reflection to invoke the methods.

**Comment by ufcpp** [comment]
[Structural typing](https://github.com/dotnet/csharplang/discussions/2548)?
[Roles and extensions](https://github.com/dotnet/csharplang/issues/5497)?

**Comment by paulhasell** [comment]
Hi,

Thanks for responding 🙂

Yes, it sounds like (another) variant of Structural Typing (which is what TypeScript is effectively applying in the same circumstances) although in my scenario where I am using the explicit implementation of an interface to select the classes I need, implicit implementation would lead to false positives.

The lightweight typing in Roles and Extensions is also similar but again would lead to implicit rather than explicit implementation.


Many Regards




________________________________


PAUL HASELL
DYNAMICS 365 PRODUCT DEVELOPMENT LEAD
m-hance
Tel: 0044 161 406 2335
Mobile: 0044 7917 899 007
Email: ***@***.***
www.m-hance.com



Disclaimer: To read our email disclaimer please visit our website.
From: Nobuyuki Iwanaga ***@***.***>
Sent: 15 December 2021 01:36
To: dotnet/csharplang ***@***.***>
Cc: Paul Hasell ***@***.***>; Author ***@***.***>
Subject: [EXT]: Re: [dotnet/csharplang] Interface equivalence (Discussion #5540)


Caution: External email

This e-mail originated outside m-hance, unless you recognise the sender and know the content is safe, do not click links or open attachments.



Structural typing<https://github.com/dotnet/csharplang/discussions/2548>?
Roles and extensions<https://github.com/dotnet/csharplang/issues/5497>?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/dotnet/csharplang/discussions/5540#discussioncomment-1809869>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE3ZBFSRO4WDU3UN33V7473UQ7WKFANCNFSM5KBLIYMA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
